lib/filewatch/discoverer.rb

```diff
@@ -0,0 +1,96 @@
+# encoding: utf-8
+require "logstash/util/loggable"
+
+module FileWatch
+  class Discoverer
+    # given a path or glob will prepare for and discover files to watch
+    # if they are not excluded or ignorable
+    # they are added to the watched_files collection and
+    # associated with a sincedb entry if one can be found
+    include LogStash::Util::Loggable
+
+    def initialize(watched_files_collection, sincedb_collection, settings)
+      @watching = []
+      @exclude = []
+      @watched_files_collection = watched_files_collection
+      @sincedb_collection = sincedb_collection
+      @settings = settings
+      @settings.exclude.each { |p| @exclude << p }
+    end
+
+    def add_path(path)
+      return if @watching.member?(path)
```

Comment:
this will likely fail if paths are not normalized, is that a consideration here?

Reply:
let me rephrase that: this will not catch duplicate paths unless they are normalized, is that a consideration here?

Reply:
👍

Reply:
I looked at this again. The path is either a file or a glob, if the glob is functionally the same as one we have seen before but not the same string then it will be added but at discovery any already discovered files will not be added to the collection - so a small overhead.

I checked the filewatch repo. This has been so since Aug 30, 2011. I don't recall any issues with this.